### PR TITLE
local stale check for session, pass through on NULL for validate

### PR DESCRIPTION
**Pull request by jread-usgs**
for #70 and #69 

@lawinslow and @aappling-usgs thoughts?

### CODE REVIEW

File: R/authenticate_sb.R [hunk 1]
```diff
@@ -47,6 +47,7 @@ authenticate_sb = function(username, password){
 		stop('Unable to authenticate to SB. Check username and password')
 	}
 	
+	attributes(h) <- c(attributes(h), list(birthdate=Sys.time()))
```
Comment (aappling-usgs) [inline R/authenticate_sb.R]: ooh, i like the storage as attribute

Comment (jread-usgs) [inline R/authenticate_sb.R]: yeah, that was one way I could avoid breaking API for session pass through stuff.

File: R/session_age.R [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	
+}
+#' @export
+#' @keywords internal
+session_expired <- function(session = current_session()){
+	
+	if(is.null(session)){
+		return(FALSE)
+	}
+	return(ifelse(session_age(session) > pkg.env$session_expires, TRUE, FALSE))
```
Comment (aappling-usgs) [inline R/session_age.R]: why not just 
```r
session_age(session) > pkg.env$session_expires
```
or
```r
isTRUE(session_age(session) > pkg.env$session_expires)
```
?

File: R/session_check_reauth.R [hunk 3]
```diff
@@ -50,7 +50,8 @@ session_check_reauth = function(username, password, session, ...){
 	}else{ #if we have passed nothing, use cached session
```
Comment (aappling-usgs) [inline R/session_check_reauth.R]: at https://github.com/jread-usgs/sbtools/blob/master/R/session_check_reauth.R#L34, reauthentication only happens if the user name has changed. what about if it expires? would it make sense to add a session_validate to that if block?

File: tests/testthat/test-eg.R [hunk 4]
```diff
@@ -0,0 +1,12 @@
+context("authentication")
+
+test_that("session validation works appropriately", {
```
Comment (aappling-usgs) [inline tests/testthat/test-eg.R]: :+1:

File: R/session_age.R (hunk 2 above)
Comment (jread-usgs) [inline R/session_age.R]: no idea why I did that. changing, thanks

File: R/session_check_reauth.R (hunk 3 above)
Comment (jread-usgs) [inline R/session_check_reauth.R]: probably, yes. session_check_reauth is confusing to me now, so I am wondering how to deal with it. It probably needs to pass the "ask" stuff we talked about.

**Comment by jread-usgs** [comment]
I've got all the POSTs covered now with internal function `sbtools_POST`, which handles validating the returned content and making sure user is authenticated.

**Comment by jread-usgs** [comment]
this is a weird one: https://travis-ci.org/USGS-R/sbtools/builds/66544990#L2595

**Comment by lawinslow** [comment]
Ha, interesting. My bet is that's due to httr automatically trying to parse the return value. Normally, it's JSON, but on error, it gets HTML. That's my theory anyway.

**Comment by jread-usgs** [comment]
yeah, looks like httr imports `jsonlite`, but only suggests `XML`

**Comment by jread-usgs** [comment]
#73 covered here for the sbtools_ GET/POST functions

**Comment by lawinslow** [comment]
@aappling-usgs go ahead and push in your fixes now.

**Pull request merged**